src/main/java/io/jenkins/plugins/analysis/core/steps/PublishIssuesStep.java

```diff
@@ -572,7 +576,13 @@ protected ResultAction run() throws IOException, InterruptedException, IllegalSt
             IssuesPublisher publisher = new IssuesPublisher(getRun(), report, healthDescriptor, qualityGate,
                     name, referenceJobName, ignoreQualityGate, ignoreFailedBuilds,
                     getCharset(sourceCodeEncoding), getLogger(report));
-            return publisher.attachAction();
+            ResultAction resultAction = publisher.attachAction();
+            QualityGateStatus status = resultAction.getResult().getQualityGateStatus();
+            if (status == QualityGateStatus.WARNING) {
```

Comment:
Note that this would only improve the visualization if the status is `WARNING`. As far as I can tell, if the status is `FAILED`, then the stage containing the step won't be visualized as failing by Blue Ocean currently, which also seems undesirable. Is there any particular reason why the step doesn't throw an `AbortException` if it sets the build result to `Failure` to abort execution and mark the source of the failure?

Reply:
I don't throw an `AbortException` since there are other post build steps running afterwards (or static analysis results collected, see https://github.com/jenkinsci/warnings-ng-plugin/blob/master/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java#L535). 

Users of my plugin typically collect several results. They want to see which individual result has a status of Success, Unstable or Failure. If the first is a Failure then the rest still can be Success. So I wonder why you do not have a similar concept for failing a step? Using only the Unstable status seems to be only half of the wanted solution. Visualization in blue ocean is also not really relevant here, since most of my users use the classic view, since they want to see the results of the analysis.

Reply:
See also:

- https://issues.jenkins-ci.org/browse/JENKINS-54373
- https://issues.jenkins-ci.org/browse/JENKINS-56288
- https://issues.jenkins-ci.org/browse/JENKINS-43995

Reply:
>  So I wonder why you do not have a similar concept for failing a step?

In a sense that is what I am trying to add via `WarningAction` to address JENKINS-39203/JENKINS-43995 - Some way to mark a step as having an error even if it didn't _really_ have an error, because right now steps only fail if they return an exception via `StepContext.onFailure`. If that doesn't happen, then as far as visualizations are concerned the step succeeded.

In https://github.com/jenkinsci/workflow-api-plugin/pull/91 I wrote:

> In the future, if we wanted to add the ability to track some new kind of non-fatal result, we could update WarningAction to store a custom Result (or similar) and use that in visualizations to modify how the action should be displayed.

I could go ahead and add that functionality so that we could also handle Failure with `WarningAction`, which would address [JENKINS-54373](https://issues.jenkins-ci.org/browse/JENKINS-54373). My only hesitation is that it seems like failures as reported by this step are exactly the use case intended to be handled by the Unstable result, but I guess it's too late to change that and if the current behavior is what users want then the _intended_ use cases for the Unstable/Failure results are kind of irrelevant.

> Visualization in blue ocean is also not really relevant here, since most of my users use the classic view, since they want to see the results of the analysis.

Fair enough. For what it's worth this would also fix the visualization in the Pipeline Steps view (see https://github.com/jenkinsci/workflow-api-plugin/pull/91#discussion_r277780826) and it would allow the visualization to be improved for Stage View (although I do not currently have plans to fix that myself).

Reply:
> I could go ahead and add that functionality so that we could also handle Failure with WarningAction, which would address JENKINS-54373. My only hesitation is that it seems like failures as reported by this step are exactly the use case intended to be handled by the Unstable result, but I guess it's too late to change that and if the current behavior is what users want then the intended use cases for the Unstable/Failure results are kind of irrelevant.

I'm not sure if I understand correctly (I'm coming from the Freestyle/UI side and use pipelines not very often). If I have several (parallel) stages that e.g. compile for different platforms then I would like to see for each stage the individual result of (Success, Unstable, Failure). Even if I get a compile error in one stage I want to compile the rest and would like to see the successful results as well. This should be similar to maven multi module builds. Even if there is one failed module I might be still interested in the actual results of the other modules (option `--fail-at-end`). 

BTW:  I would only use Unstable for analysis results by myself. Just several of my users requested to use the Failure status as well for the quality gate.

Reply:
> If I have several (parallel) stages that e.g. compile for different platforms then I would like to see for each stage the individual result of (Success, Unstable, Failure).

Yes, `WarningAction` as it exists now would make this work for `Success` and `Unstable`, but would need some additions to work for `Failure`. I am happy to make those additions if you think it would be useful.

> Even if I get a compile error in one stage I want to compile the rest and would like to see the successful results as well. This should be similar to maven multi module builds. Even if there is one failed module I might be still interested in the actual results of the other modules (option --fail-at-end)

The default behavior of the `parallel` step already matches `--fail-at-end`: It waits until all branches complete before moving on, so even if one branch fails they would all execute to completion (and if users want the parallel step to fail immediately if a branch fails there is an option to do so). If you run analyses sequentially, then you do need something like `WarningAction` to keep the build from stopping at the first failure while still being able to see which step failed.